fix: hide topbar for fullscreened floating and scratch windows (fixes #451)

jtaala opened the pull request:
Fixes #451.

Checks if the window is fullscreen and on the scratch layer (or is fullscreen and floating), if so it hides the topbar.

jtaala committed:
Added check to fixTopBar() to check if focused window is on scratch and
fullscreen (hide topbar).

hallettj reviewed (dismissed):
I have been seeing this issue in a game running under Wine. I can confirm that this change fixes the problem. Once again I appreciate your fix!

jtaala commented:
Morning all! @Gelbana @tadfisher @ccope @smichel17  - here's another quick fix to review (if needed, noting that @hallettj approved this this PR previously - thanks @hallettj!) when someone has time.

Simply also hides the top bar when a window on the scratch layer (or a floating window) goes fullscreen.

smichel17 commented on topbar.js:
```diff
@@ -586,6 +586,11 @@ function fixTopBar() {
     let fullscreen = selected && selected.fullscreen && !(focusIsScratch);
     let hideTopBar = !spaces.monitors.get(panelMonitor).showTopBar
+    // hide topbar for fullscreen scratch windows
+    if (focusIsScratch && focus.fullscreen) {
```
One particular thing I would look for: `focusIsScratch` currently includes a check to make sure that `focus` is not null, so it's safe to check `focus.fullscreen` here. But there's no guarantee `focusIsScratch` will keep that check, so it would be easy to end up with a crash here from trying to access a property of null/undefined.

However this ends up, I'd either try to get it to match the `if (!focus) return` pattern so it's safe to use in the rest of the function, or pull things out so there's an explicit `focus && focus.fullscreen` somewhere or other.

jtaala replied:
yes, good call tbh - it was/is quite messy and a refactor is probably in order!

I'll relook at the implementation of this and go for a cleaner approach/refactor.

smichel17 replied:
I think I'd start with something like this (if I'm guessing correctly about what `selected` vs `focus` means) and go from there. It doesn't fully address the null check issue, but it moved them closer together at least, and I think would simplify the logic below.
````diff
-let fullscreen = selected && selected.fullscreen && !(focusIsScratch);
+let fullscreen = focusIsScratch ? focus.fullscreen : selected && selected.fullscreen;
````

jtaala replied:
Thanks @smichel17, so, I've simplified the current fixTopBar() function as a first step (see current commit) - and added comments detailing (my) current understading/interpretation of what is happening in the code.

Let me know if that aligns with what you're seeing/thinking (especially the diff between `selected` and `focused` window).

Using your suggested logic for `fullscreen` boolean (looks good to me!).

I've been testing this simplification (no issues so far for both tiled/scratch fullscreen windows) and will continue testing this.

jtaala replied:
Thanks @smichel17 - please checkout the latest refactor.  Requesting a re-review.  Cheers!
edit: yes, chain of if/else's makes sense here and improves readability and flow.

smichel17 reviewed (changes requested):
This may work, but it makes already messy logic even messier.

Why do we exclude scratch windows from our `fullscreen` check variable, but then immediately check if a window is scratch and fullscreen? Why do we have a `hideTopBar` variable but then it's only one of three situations where we hide the top bar?

I think some combination of changing what goes into the intermediate variables and naming them more descriptively would do wonders here. Also (this one's a personal preference), I would swap the early `return`s at the end for a chain of `else if`s with the last two lines of `fixTopBar` going in the closing `else` (~~if there are still 3 cases post-refactor, at least~~ edit: actually I just strictly prefer it; I'd do it even with a single if/else).

jtaala committed:
Merge remote-tracking branch 'upstream/develop' into fix-issue451-hide-topbar-on-scratch-fullscreen

jtaala committed:
Cleaned up and simplified - added comments on detail current
understanding on what is happening here (before finalise refactor).

jtaala committed:
Small improvementment in readability.

jtaala committed:
Added case not addressed: floating windows (different to scratch windows).
Floating windows should have topbar hide on fullscreen too.  Floating windows are not scratch windows.
An example is the PaperWM extension settings - it's floating but not a scratch window.

jtaala commented:
Huzzah! found a case not addressed: _floating windows_.  Floating windows are not scratch windows but should also have the topbar hide (floating windows can be fullscreen'd but their top is then not visible since the topbar is covering them).  Latest commit addresses these windows too.

An example of a floating window is the PaperWM extension settings - it opens and floats but is not on the scratch layer.

jtaala committed:
Small simplification and comment improvement.

jtaala commented on topbar.js:
```diff
@@ -574,30 +574,35 @@ function disable() {
-    if (!spaces)
-        return;
-    let space = spaces.monitors.get(panelMonitor);
+    let space = Tiling.spaces?.monitors.get(panelMonitor) ?? false;
```
Just to explain this line (in case others aren't use to such js operators) - it's using an optional chain with the nullish coalescing operator `??`.

Basically, if `Tiling.spaces` is null (or undefined) then the nullish coalescing operator (`??`) means `space` will be assigned to `false`; otherwise it will return the left hand side of `??`, with the optional `?.` function applied, meaning `space` will then be assigned the value of `Tiling.spaces.monitors.get(panelMonitor)` (since `Tiling.spaces` is not `null` and not `undefined`).

jtaala commented:
@Gelbana @tadfisher @ccope @smichel17 @hallettj - any further comments on this one?  I believe it's ready to go and the rewrite is much clearer as to the intent and function of `fixTopBar` (plus it now catches, I believe, all the cases - e.g. if a window is fullscreen'd AND (tiled or floating,or on the scratch layer).

smichel17 reviewed (dismissed):
Maybe a little overcommented now (main useful distinction was focused vs selected, since the latter is a paperwm-specific concept that I didn't understand at first). But looks good.

jtaala commented:
> Maybe a little overcommented now (main useful distinction was focused vs selected, since the latter is a paperwm-specific concept that I didn't understand at first). But looks good.

Lol - I actually read through it again yesterday and thought the same thing!  

I'll clean up the comments (not sure it it would need to be approved again though - don't want to make more work for you!).

jtaala committed:
Clean up of superfluous comments.
